devito/yask/transformer.py

```diff
@@ -0,0 +1,178 @@
+from devito.dimension import LoweredDimension
+from devito.ir.iet import FindNodes, Expression
+from devito.ir.support import Backward, Forward
+from devito.logger import yask_warning as warning
+from devito.symbolics import split_affine
+
+from devito.yask import nfac
+
+__all__ = ['yaskizer']
+
+
+def yaskizer(trees, yc_soln):
+    """
+    Populate a YASK compiler solution with the :class:`Expression`s found in an IET.
+
+    The necessary YASK grids are instantiated.
+
+    :param trees: A sequence of offloadable :class:`IterationTree`s, in which the
+                  Expressions are searched.
+    :param yc_soln: The YASK compiler solution to be populated.
+    """
+    # Track all created YASK grids
+    mapper = {}
+
+    # It's up to Devito to organize the equations into a flow graph
+    yc_soln.set_dependency_checker_enabled(False)
+
+    processed = []
+    for tree in trees:
+        # All expressions within `tree`
+        expressions = [i.expr for i in FindNodes(Expression).visit(tree.root)]
+
+        # Attach conditional expression for sub-domains
+        conditions = [(i, []) for i in expressions]
+        for i in tree:
+            if not i.dim.is_Sub:
+                continue
+
+            ydim = nfac.new_domain_index(i.dim.parent.name)
+
+            try:
+                # Are the extremes statically known ?
+                lower = nfac.new_const_number_node(int(i.dim.symbolic_ofs_lower))
+                upper = nfac.new_const_number_node(int(i.dim.symbolic_ofs_upper))
```

Comment:
This obviously works, but it _looks_ like it might ignore some possible cases.
YASK allows you to construct any expression of the form FIRST+integer or LAST+integer, right?

So there are three cases, I imagine:

1. SubDimension goes from x_m+3 to x_M-3 (handled correctly)
2. SubDimension goes from x_m to x_m+3 (left boundary - I think not handled by this code since symbolic_ofs_upper is not an integer).  This could be expressed in the YASK API, though, as far as I can see.
3. SubDimension goes from x_M-3 to x_M (same situation for the right boundary - symbolic_ofs_lower not an integer)

But it passes the tests, so I'm not sure what I'm missing here.

Reply:
>  I think not handled by this code since symbolic_ofs_upper is not an integer)

but then you fall back to the ``except`` below, don't you? where ``i.extent()`` is read... This should cover cases 2 and 3.
What am I missing?

Reply:
The `except:` part only applies to sequential loops, though.  So that needlessly excludes boundary conditions or other regions where some parallelism is possible - i.e. those without the intra-timestep dependency like the one that we are using.

Something like:

```
start_expressible_in_yask = (
    (i.dim.symbolic_start - i.dim.parent.symbolic_start).is_Number
    or (i.dim.symbolic_start - i.dim.parent.symbolic_end).is_Number)
end_expressible_in_yask = (
    (i.dim.symbolic_end - i.dim.parent.symbolic_start).is_Number
    or (i.dim.symbolic_end - i.dim.parent.symbolic_end).is_Number)
subdim_statically_known = start_expressible_in_yask and end_expressible_in_yask
```

Now that I think about it, is it also possible that you might need to check for is_Parallel in the top part?

Reply:
uhm, I think I'm confused and have to think about this !

Reply:
do you suggest to enrich the test suite with some further tests? because if you spotted this, assuming it's actually a bug/issue, then we will have to cover it anyway. 
Of course, no rush: at this point, we can leave this open until next week. But if you could produce one or two MFEs, that'd be great!

Reply:
Just added some examples to `test_yask.py`.  

I think the logic needs to be something like:

* if the loop is parallel AND bounds can be expressed as a yask condition, then
   - build the yask condition, and apply it
* otherwise, if the loop is sequential, and the number of iterations is known at compile time, then:
   - unwind into a series of yask regions, each with its own condition, and with dependencies to enforce the order
* otherwise, don't offload to yask

Is that sensible?

Reply:
> Just added some examples to test_yask.py.

These seem very relevant for ``core`` too, so I hope you don't mind if I move them to the more generic ``test_dimension.py``? Anyway, these look great !

> I think the logic needs to be something like [...] Is that sensible?

yeah... I mean, there are things of the current implementation that I can't remember why I did them this way 😮 as in, the (flow)[https://github.com/opesci/devito/blob/yask-regions/devito/yask/transformer.py#L41] kinda matches what you described, but there are several things that I don't quite understand myself ! Hopefully I'll be able to elucidate soon

Reply:
Also note that one of the tests: `test_yask.py::TestOperatorSimple::test_subdimleft_parallel` passes, but it hasn't offloaded to YASK as it should.  I couldn't immediately see an example of how to test that the offload worked.

Reply:
a hacky way is to check for the presence of the string `run_solution` within `str(op)`

Reply:
> So that needlessly excludes boundary conditions or other regions where some parallelism is possible - i.e. those without the intra-timestep dependency like the one that we are using.

This should be covered in the ``try`` part. Did  you see what ``i.dim.symbolic_ofs_lower/upper`` return?

Reply:
I can't see how, though...?

```
 def symbolic_ofs_upper(self):
        return self._interval.right - self.parent.symbolic_end
```

For a `SubDimension.left`, we might have `interval.right = x_m + 3`.  So the above would return `x_m + 3 - x_M`, which ought to fail to cast to `int`...?